### DISCUSSION TITLE
Large, single process vs many nested processes

### DISCUSSION THREAD

**Discussion by jacobpennington** [lava-nc/lava#460] [General]
(Continuing from questions during the conference today)

My understanding from developer responses today is that Lava (and possibly Loihi itself?) work best when all neurons of a given type are managed by a single process. The existing LIF neuron is coded this way: for a group of 100 neurons, you use a single LIF process with shape (100, ) for states and ports (and a corresponding process model describes the behavior of updating all 100 neurons simultaneously).

However, from a development (and general mathematical) standpoint I find it's often easier to implement new models at a more granular level. For simple LIF dynamics this isn't really necessary, but just to use it as a first example:   I would like to instead program a process named, perhaps, LIF_group that is a simple wrapper around some number of LIF subprocesses, where each subprocess has shape (1,). Then the process model for LIF only needs to worry about single-neuron dynamics, with the drawback being that the process model for LIF_group would now be more complicated since it has to coordinate communication between many ports, or possibly other issues that I'm not considering.

For more complicated neuron models, the latter drawback might be well worth it to simplify implementing the dynamics. For instance, our group is interested in implementing generalized leaky integrate and fire (GLIF) models from this paper:
https://www.nature.com/articles/s41467-017-02717-4
Which introduces several additional state variables for post-spike currents, spike-dependent threshold, etc. I'll give a concrete example of trouble that I encountered when sketching out Lava code for one version of this (GLIF3, which just adds the post-spike currents) as a single process for many neurons. 

For the post-spike currents in this model, the value of each of two currents would be added to the membrane potential at each time step for each neuron (with some other details omitted). For the CPU implementation it's most natural to me to do this as:
```python
# For voltage shape (N, )  and after_spike_currents shape (N, 2)
voltage[...] += np.sum(after_spike_currents, axis=1)
```
But this wouldn't be available when implementing the Loihi 2 process model. Of course the same thing can be achieved with for loops, but the even simpler thing would be implementing from a single-neuron perspective to begin with:
```python
# For all scalar voltage, current1, current2
voltage += (current1 + current2)
```
which I can be much more confident will translate readily to a Loihi 2 implementation.

In short, the issue isn't that implementing the models isn't possible with the current paradigm, or even that it's "hard" to work around it for this specific model, but it would be more convenient/intuitive (possibly much moreso for other models) if I could implement process models for neuron dynamics at the single-neuron level (without, apparently, incurring performance penalties or running into hardware limitations).

**Comment by awintel** [comment]
Thanks for your feedback.

Let me first clarify on some context because I wouldn't make this statement: "work best when all neurons of a given type are managed by a single process".
There is absolutely no requirement that Loihi (Leaf)Procs have to contain a homogeneous set of neurons. A multi-neuron Process may very well contain different types of neurons without performance implications either on CPU or Loihi. The only reason the current standard library LIF Process is homogeneous is because most of the time we just need a homogenous array of neurons of the same type and because it allows for a cleaner more focused user interface.
However, there are different implications for splitting an N dimensional Process into more less than N dimensional Processes but for different reasons on Loihi and CPU.
- CPU: Since Lava has an asynchronous and concurrent execution model, every e.g., Python Process and Channel require their own thread. This immediately implies that fewer of them are preferable to avoid an explosion of threads if you put every neuron and its 1-d output channel in their own thread. In addition, executing vectorized e.g., numpy code on CPU/GPU is much more efficient motivating to combine the math of many similar neurons rather than looping over single neurons. This isn't even specific to Lava.
- Loihi: Right now, the compiler begins a new set of cores for each new neural Process and all its associated input connection Processes that need to be compiled together into the same NeuroCores. Thus, having many tiny Processes would lead to a lot of underutilized cores. This could eventually be relaxed but makes compilation a little less efficient. This simplification was made based on the observation that most of the time, Processes are homogeneous and span multiple cores anyways.

Can you clarify what you mean by "But this wouldn't be available when implementing the Loihi 2 process model"? It seems you have a population of N identical neurons but with each having 2 inputs thus leading to an InPort of size 2*N. This is perfectly legal by writing your own micro code implementation. It can be derived from an LIF but it's not a pure LIF model anymore. Also having Processes for single LIF neurons wouldn't help because you want to combine internal state variables of two LIF neurons. It's just a slightly different neuron model.

If you would like to share your code eventually with the community or would like specific feedback on ProcModels from our team, then you can also point us to your fork where you develop your code or open a PR eventually.

Thanks

**Reply by jacobpennington** [reply]
> There is absolutely no requirement that Loihi (Leaf)Procs have to contain a homogeneous set of neurons.

My apologies, I didn't intend to imply this. Good to know though.

> In addition, executing vectorized e.g., numpy code on CPU/GPU is much more efficient motivating to combine the math of many similar neurons rather than looping over single neurons. This isn't even specific to Lava.

Of course. My suggestion was more along the lines of "user specifies dynamics for a single neuron, Lava objects figure out how to run that as vectorized code behind the scene" rather than requiring users to implement both dynamics and vectorization for new models. Of course that puts a lot more burden on the Lava development team, so maybe it's not feasible. However, after seeing the Brian2Lava demonstration on the second day I think that will more or less fill this niche anyway.

> Can you clarify what you mean by "But this wouldn't be available when implementing the Loihi 2 process model"?

I was referring specifically to the numpy code, not the model in general. I.e. I can write "a = b + c" essentially the same way in both microcode and python, but "np.sum(..., axis=1)" does not have an equivalent microcode operation as far as I understand so the implementation would be different (I've not used microcode before, so I may be mistaken here). Again, I'm not claiming that such a model *can't* be implemented in Loihi 2, it definitely looks like it can be. I'm just making a suggestion regarding ease-of-coding in the context of coding up process models for multiple backends.

> If you would like to share your code eventually with the community or would like specific feedback on ProcModels from our team, then you can also point us to your fork where you develop your code or open a PR eventually.

I'll be sure to do that in the future. At the moment it's just notes and rough prototypes and then I had to switch priorities, not worth having others look through yet.

Thanks for your replies!
